Feat/analytics

JuroOravec opened the pull request:
- Introduces analytics store module.
- Introduces analytics Vue plugin with auto page tracking and directive `analytics-model` to trigger analytics store module actions on `v-model` change.
- Replaces VueAnalytics with [analytics](https://github.com/DavidWells/analytics). GA tracking implemented via [analytics ga plugin](https://getanalytics.io/plugins/google-analytics).
- Introduces `a-extended` component for link tracking.
- Tracking is toggable via store and via env vars (same as consents, off by default, and on on staging and prod).
- Tracking is disabled on `DoNotTrack` browser setting, when analytics are disabled, or when `"statistics"` cookie consent is not given.
- Introduces bunch of analytics plugins for enriching/modifying the payload, and bunch of helper functions to add/remove these plugins, in `src/utils/analytics`
- Introduces packages `analytics`, `analytics-plugin-do-not-track`, `analytics-plugin-ga`, `analytics-plugin-original-source`
- Tracks following events:
  - `search_escher` - when user searches escher (and the keyword doesn't change for 2 sec)
  - `send_to_visualize` - when predictions are sent for visualization
  - `visualize` - when Escher map is being (re)loaded. Describes which card/map/project/etc is being visualized.
  - `export` - when user exports item(s)
  - `link` - when user clicks on an external link (external link is defined as having different hostname)
  - `track_dialog` - when a new dialog is open (new model, new project, new map, card, ...)
  - `page` - when user navigates to different page
  - `login` - when user logs in (either via social/email login or from localStorage)
-  on registration or login (either via social/email login or from localStorage), user is identified by the following (depending on availability): internal user id, firebase id, or hashed email
- on registration or login (either via social/email login or from localStorage), user info is captured (depending on availability)
- on logout, user is forgotten and is treated as an anonymous user.

JuroOravec committed:
feat: add base analytics module to vuex store

JuroOravec committed:
feat: identify user on login

Identifies at email login, social login and registration.
At registration, the params are emitted to LoginDialog where
the identification happens. User is identified by 3 IDs, each of which
should be unique to the given user:
registeredEmail, internalId and firebaseId

JuroOravec committed:
feat: update user props on login or registration

JuroOravec committed:
feat: capture login event on email or social login

Login event captures the login method (e.g. "email", "github", etc),
and whether this is user's first login (== registration).

JuroOravec committed:
feat: track logging in for from localStorage

JuroOravec committed:
feat: capture logout event

JuroOravec committed:
feat: enable link tracking

JuroOravec committed:
feat: add a-extended component

a-extended behaves like regular <a>, but upon click it checks if href points to
remote address, and dispatches "analytics/link" with destination info
if so.

JuroOravec committed:
refactor: replace <a> with <a-extended>

JuroOravec committed:
feat: capture "visualize" events

Tracks "visualize" event at every point where the modification/design
data are changed for current card
(incl. changes caused from model manipulation from within Escher),
or initial/different map or card is selected, such that the Escher map
is (re)calculated.

JuroOravec committed:
feat: capture "export" events

Tracks "export" event on job results data export

JuroOravec committed:
feat: capture "searchEscher" event

Tracks "searchEscher" event on search in Escher

JuroOravec committed:
chore: update escher to support search callback

JuroOravec committed:
fix: AIMUser interface is not defined missing

The interface represents the data returned from the GET request
on iam/user endpoint

JuroOravec committed:
fix: analytics/loginLocalStorage missing imports

JuroOravec committed:
chore: add analytics package

JuroOravec committed:
feat: add analytics plugin

The analytics plugin wraps together analytics package, and
page-tracking bits from vue-analytics package.

JuroOravec committed:
feat: add v-model-bound analytics directive

The directive calls one of commands passed to analytics plugin when the
component's v-model changes to truthy value

JuroOravec committed:
feat: set up dialog tracking

Apply v-analytics-model on dialogs throughout the app.

JuroOravec committed:
fix: incorrect closing tag

JuroOravec committed:
fix: no payload passed through v-analytics-model

JuroOravec committed:
chore: add license to vue-analytics files

JuroOravec committed:
feat: add analytics/page action

This action is called by page tracking and dialog tracking.

JuroOravec committed:
feat: capture "sendToVisualize" events

Occurs when designs or predictions are selected for visualization of
interactive map. Note that this is different than the "visualize" event,
which tracks when a single card/design is being rendered in Escher

JuroOravec committed:
refactor: track search in Escher only after delay

JuroOravec committed:
refactor: v-analytics-model calls analytics action

Rather than configuring object of commands, the command name passed to
the v-analytics-model directive will call respective action from
"analytics" vuex store module only.
This couples the plugin to the analytics store module, but for current
use it's imo more appropriate than handling additional object of
functions.

JuroOravec committed:
feat: add "trackDialog" action

Analytics module action endpoint for capturing when a dialog has been
opened by a user.

JuroOravec committed:
feat: add support for analytics toggle-ablility from env vars

JuroOravec committed:
feat: replace VueAnalytics with analytics package

JuroOravec committed:
feat: add analytics plugin to disable analytics

JuroOravec committed:
feat: add analytics plugin to disable analytics on missing consent

JuroOravec committed:
feat: add analytics plugin to disable analytics on DoNotTrack

JuroOravec committed:
feat: add originalSourcePlugin to analytics

JuroOravec committed:
fix: compiler error - no default destruct values

JuroOravec committed:
chore: add vue analytics plugin config type

JuroOravec committed:
feat: add helper function for chaining custom plugins

Enables stepwise changes to the payload and easy addition/removal/order change of plugins.

JuroOravec committed:
feat: payload-modifying hook factory

To be able to easily create functions which modify part of payload and
can be hooked to `track`, `identify` or `page` events.

JuroOravec committed:
feat: add analytics plugin to enrich tracked data

JuroOravec committed:
feat: add analytics plugin to drop nil values

JuroOravec committed:
feat: add analytics plugin to convert camelCase properties to snake_case

JuroOravec committed:
feat: add analytics plugin to print out payload

Used for debugging purposes.

JuroOravec committed:
fix: dropNoValuePropertiesPlugin missing imports

JuroOravec committed:
fix: namespacePluginHooks is not exported

JuroOravec committed:
style: fix TS compiler errors

JuroOravec committed:
refactor: make consent plugin hooks in analytics async

To ensure the functions are ran only after consents have been loaded.

JuroOravec committed:
feat: add info from analytics originalSourcePlugin to enrichAnalyticsPlugin

JuroOravec committed:
refactor: use snakeCasePropertyNames from utils

JuroOravec committed:
refactor: rename and modify makePayloadPropertyModifierFactory

Renamed to makePayloadPropertyMutatorFactory.
Refactored to support both normal and namespaced hooks.
Refactored to pass only config to the mutator function instead of the
whole plugin object.

JuroOravec committed:
fix: wrong func in namespacePluginHooks err msg

JuroOravec committed:
refactor: chain plugins via wrapping instead of namespaced hooks

Namespaced hooks for analytics package plugins did not compound the
changes to the payload object. So instead create a single plugin that
does the changes but is composable.

JuroOravec committed:
feat: catch errs from 3rd party analytics plugins

JuroOravec committed:
fix: function capitaze not imported

JuroOravec committed:
refactor: support async in makePayloadPropertyMutatorFactory

JuroOravec committed:
feat: add hashMessage to utils/analytics

JuroOravec committed:
refactor: hash user ID in analytics

JuroOravec committed:
fix: split error when __user_original_source not in localStorage

JuroOravec committed:
fix: missing deps and typos in package-lock.json

kvikshaug commented on src/utils/utility.ts:
```diff
@@ -104,3 +104,11 @@ export function camelCasePropertyNames(obj) {
+  } catch (e) {
+    console.warn(e);
+  }
+}
```
This is an odd function, I recommend silencing explicitly where needed.

JuroOravec replied:
I removed it. Originally, it was meant to silence GA analytics plugin error on invalid trackingID, but it turned out that the plugins are initialized asynchronously, so no need for the try/catching.

kvikshaug reviewed (changes requested):
- The analytics package [seems to be developed by a single contributor](https://github.com/DavidWells/analytics/graphs/contributors), so the project has a high bus-factor. I'm also not seeing a huge community around this dependency so I recommend to reconsider using it.
- The plugins are a huge introduction of low-level code and I am worried about maintainability. Isn't it possible to bring these in as dependencies without adding the code in our repo?
- I am not sure we can introduce MIT licensed code in our repository as all our code needs to be open sourced under the Apache license.
- I'm a bit on the fence about the `a-extended` tag, I think it's a good abstraction but it can be hard for new developers to realize they need to use it over a regular anchor in order to track a click. Maybe documentation would solve that.

As for the Travis build, you need to run the linter, and it also looks like there are a number of typescript issues that need to be resolved.

JuroOravec committed:
refactor: remove console statements

To resolve errors at compilation checks

JuroOravec committed:
style: linting

JuroOravec committed:
refactor: Resolve type issues

JuroOravec committed:
refactor: remove tryCatchCall

JuroOravec commented:
> * The analytics package [seems to be developed by a single contributor](https://github.com/DavidWells/analytics/graphs/contributors), so the project has a high bus-factor. I'm also not seeing a huge community around this dependency so I recommend to reconsider using it.

It's true about the single contributor, but the use also seems to be growing steadily since Oct 2019 where it seems to have got some recognition or was put into use. The interaction with that package is limited to the inside of the analytics store module, and the package itself depends only on redux, so I think it's fairly safe. So I suggest to use it, and if push comes to shove and it really gets broken and forgotten, then replacing the calls to that package with direct calls to e.g. Google Analytics should be fairly simple.

> * The plugins are a huge introduction of low-level code and I am worried about maintainability. Isn't it possible to bring these in as dependencies without adding the code in our repo?
> * I am not sure we can introduce MIT licensed code in our repository as all our code needs to be open sourced under the Apache license.

Hm, how exactly does licensing work? I thought there's no difference if you use a package as a dependency to your source code vs when you have the actual code in your code base? 

So do I understand correctly that you're saying that if the MIT-licensed code lived on its own, its distribution could be used as a dependency without issues?

Also, about the maintainability, are you pointing at the code that manages/creates the plugins, or with the plugins included?  I'm asking because half of the custom plugins are specific to our use case (`requireConsentPlugin` and `disableAnalyticsPlugin`, `enrichAnalyticsPlugin`), so in their case it'd make more sense to do those actions on the payload inside the store before it's passed to the analytics than have them as dependencies. For others, like `dropNoValuePropertiesPlugin` `snakecasePropertiesPlugin`, these are general enough that they could be separate.

So if I understand correctly, it would be desirable to extract the code that encapsulates `VueAnalytics` and `analytics` packages, along with other related helper functions, and have that as a separate dependency? And then also package separately `dropNoValuePropertiesPlugin` and `snakecasePropertiesPlugin`?

> * I'm a bit on the fence about the `a-extended` tag, I think it's a good abstraction but it can be hard for new developers to realize they need to use it over a regular anchor in order to track a click. Maybe documentation would solve that.

In the spirit of documenting up how things work, it might be actually good to have a doc folder that would explain how the frontend app is wired up, and there it could also include custom components like the `a-extended`? What do you think?

> As for the Travis build, you need to run the linter, and it also looks like there are a number of typescript issues that need to be resolved.

Yes, that was likely also the cause of the issue I was getting. Fixed the type issues and all is fine now. But it's weird that Travis reported missing dependencies as the cause of failure instead of the actual type errors.

JuroOravec committed:
refactor: trim and rewrite analytics plugins

JuroOravec committed:
refactor: resolve merge conflicts

JuroOravec committed:
refactor: scope data processing only to selected plugins

Before, the processing pipelines ran twice. Once for the
process-payload.track/identify/page methods, and once for
process-payload["track:google-analytics"] methods.

JuroOravec committed:
refactor: capture right-click on external link

JuroOravec commented:
I trimmed and rewrote the logic and replaced licensed files. The analytics package plugins, vue directive and vue plugin are now all in `src/plugins/analytics.ts`

JuroOravec committed:
feat: add analytics state to RootState

kvikshaug commented on .env.staging:
```diff
@@ -2,4 +2,5 @@ VUE_APP_ENVIRONMENT=staging
 VUE_APP_GA_TRACKING_ID=UA-106144097-3
-VUE_APP_ENABLE_CONSENTS=true
+VUE_APP_ENABLE_ANALYTICS=true
+VUE_APP_ENABLE_CONSENTS=true
```
Files should end with a newline

JuroOravec replied:
fixed

kvikshaug commented on src/utils/utility.ts:
```diff
@@ -1,4 +1,4 @@
-import Vue from "vue";
+import * as Sentry from "@sentry/browser";
```
Looks like an unused import

JuroOravec replied:
fixed. Thanks for spotting it. I don't actually remember why it was imported in the first place.

JuroOravec committed:
Merge branch 'devel' into feat/analytics

kvikshaug reviewed (approved):
Thanks for the updates!

JuroOravec committed:
style: lint

JuroOravec committed:
refactor: remove unused import

JuroOravec committed:
style: empty line at end of file
